database/sqlite/sqlite_functions.c

```diff
@@ -0,0 +1,778 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "sqlite_functions.h"
+
+#define ENABLE_CACHE_CHARTS 1
+#define ENABLE_CACHE_DIMENSIONS 1
+
+
+#define error_report(x, args...) { errno = 0; error(x, ##args); }
+
+const char *database_config[] = {
+    "PRAGMA auto_vacuum=incremental; PRAGMA synchronous=1 ; PRAGMA journal_mode=WAL; PRAGMA temp_store=MEMORY;",
+    "PRAGMA journal_size_limit=17179869184;",
+    "ATTACH ':memory:' as ram;",
+    "CREATE TABLE IF NOT EXISTS host (host_id blob PRIMARY KEY, hostname text, "
+    "registry_hostname text, update_every int, os text, timezone text, tags text);",
+    "CREATE TABLE IF NOT EXISTS chart (chart_id blob PRIMARY KEY, host_id blob, type text, id text, name text, "
+    "family text, context text, title text, unit text, plugin text, module text, priority int, update_every int, "
+    "chart_type int, memory_mode int, history_entries);",
+    "CREATE TABLE IF NOT EXISTS dimension(dim_id blob PRIMARY KEY, chart_id blob, id text, name text, "
+    "multiplier int, divisor int , algorithm int, options text);",
+
+    "CREATE TABLE IF NOT EXISTS chart_active(chart_id blob PRIMARY KEY, date_created int);",
+    "CREATE TABLE IF NOT EXISTS dimension_active(dim_id blob primary key, date_created int);",
+    "CREATE INDEX IF NOT EXISTS ind_host on chart (host_id);",
+    "CREATE INDEX IF NOT EXISTS ind_chart on dimension (chart_id);",
+
+    "delete from chart_active;",
+    "delete from dimension_active;",
+    NULL
+};
+
+sqlite3 *db = NULL;
+
+static uv_mutex_t sqlite_lookup;
+static uv_mutex_t sqlite_add_page;
+static uint32_t page_size;
+static uint32_t page_count;
+static uint32_t free_page_count;
+
+/*
+ * Database parameters
+ */
+uint32_t sqlite_disk_quota_mb; // quota specified in the database
+uint32_t desired_pages = 0;
+
+static int execute_insert(sqlite3_stmt *res)
```

Comment:
It appears we are randomly dropping "INSERT" statements if the DB is busy and the timeout is reached. This will corrupt the database. How do we fix this?

Reply:
I would say retry forever, this is what we were doing with netdata custom DBs before anyways.